File: 6-kyu-kata/error-correction-number-1-hamming-code/python/error_correction_number_1_hamming_code.py
# 6 kyu - Error correction #1 - Hamming Code  [ ID: 5ef9ca8b76be6d001d5e1c3e  (error-correction-number-1-hamming-code) ]
# URL: https://www.codewars.com/kata/5ef9ca8b76be6d001d5e1c3e
# Category: ALGORITHMS  |  Tags: ALGORITHMS | BITS
# ******************************************************************************
import re
  
  def encode(string):
      e1 = re.sub(".", (lambda y: ("0000000" + (bin(ord(y.group()))[2:]))[-8:]), string)
      return re.sub(".", lambda x: x.group() * 3, e1)
  
  def decode(bits):
      d1 = re.sub("...", lambda x: x.group()[0] if x.group()[0] in (x.group()[1], x.group()[2]) else x.group()[1], bits)
      return re.sub(".{8}", lambda y: chr(int(y.group(), base=2)), d1)

File: kata-6-kyu/error-correction-number-1-hamming-code/javascript/errorCorrectionNumber1HammingCode.js
//+ ====================================================================================================================
//+
//+ 6 kyu - Error correction #1 - Hamming Code  [ ID: 5ef9ca8b76be6d001d5e1c3e ] (error-correction-number-1-hamming-code)
//+ URL: https://www.codewars.com/kata/5ef9ca8b76be6d001d5e1c3e
//+ Category: ALGORITHMS  |  Tags: ALGORITHMS | BITS
//+
//+ ====================================================================================================================

const encode = (text) =>
  [...text]
    .map((v) =>
      `0000000${v.charCodeAt().toString(2)}`
        .slice(-8)
        .split("")
        .map((v) => v.repeat(3))
        .join("")
    )
    .join("")

const decode = (bits) =>
  bits
    .match(/.../g)
    .reduce((a, [c1, c2, c3]) => a + ([c2, c3].includes(c1) ? c1 : c2), "")
    .match(/.{8}/g)
    .map((v) => String.fromCharCode(parseInt(v, 2)))
    .join("")

module.exports = { encode, decode }
